Logging retries

nickrmc83 opened the pull request:
<!-- Thank you for contributing to osquery! -->

To submit a PR please make sure to follow the next steps:

- [ ] Read the `CONTRIBUTING.md` guide at the root of the repo.
- [ ] Ensure the code is formatted building the `format_check` target.  
      If it is not, then move the committed files to the git staging area,
      build the `format` target to format them, and then re-commit.
      [More information is available on the wiki](https://osquery.readthedocs.io/en/latest/development/building/#formatting-the-code).
- [ ] Ensure your PR contains a single logical change.
- [ ] Ensure your PR contains tests for the changes you're submitting.
- [ ] Describe your changes with as much detail as you can.
- [ ] Link any issues this PR is related to.
- [ ] Remove the text above.

<!--

The PR will be reviewed by an osquery committer.
Here are some common things we look for:

- Common utilities within `./osquery/utils` are used where appropriate (avoid reinventions).
- Modern C++ structures and patterns are used whenever possible.
- No memory or file descriptor leaks, please check all early-return and destructors.
- No explicit casting, such as `return (int)my_var`, instead use `static_cast`.
- The minimal amount of includes are used, only include what you use.
- Comments for methods, structures, and classes follow our common patterns.
- `Status` and `LOG(N)` messages do not use punctuation or contractions.
- The code mostly looks and feels similar to the existing codebase.

-->

Smjert committed:
cve: Update sqlite to 3.45.0 (#8259)

nickrmc83 committed:
Upgrade rocksdb; add more AWS logger logging

- Update rocksdb to a new version
- reduce the retries the AWS Kinesis logger will attempt
- cap the retry delay to 15 seconds down from 103!
- add more logging of failures

loranmutafov reviewed (approved):
It's your account, but still LGTM 😄

nickrmc83 committed:
Fix unsigned-ness of int, fix rocksdb

nickrmc83 committed:
Fix linking issues

nickrmc83 commented on plugins/logger/aws_kinesis.cpp:
```diff
@@ -147,7 +147,7 @@ size_t KinesisLogForwarder::getMaxBytesPerBatch() const {
 }
 
 size_t KinesisLogForwarder::getMaxRetryCount() const {
-  return 100U;
```
This retry count would end up taking a very long time because of the compounding of AWS retries (3 by default) and then increases per outer loop.

nickrmc83 committed:
Add rocksdb code, sub-code and severity to error logs

nickrmc83 committed:
Treat soft errors as successes.

nickrmc83 committed:
Add rocksdb auto-recovery logic

nickrmc83 committed:
update auto recover count to 50

nickrmc83 committed:
Comments

nickrmc83 committed:
comments and tidying

nickrmc83 commented on plugins/database/rocksdb.cpp:
```diff
@@ -365,17 +397,39 @@ Status RocksDBDatabasePlugin::putBatch(const std::string& domain,
+      return Status(Status::kSuccessCode, error_string);
+    case rocksdb::Status::Severity::kHardError:
+      LOG(ERROR) << "Hard error encountered during putBatch, continuing optimistically but this event is lost: " << error_string;
+      return Status(Status::kSuccessCode, error_string);
```
An error should be returned here and the caller adjusted to accept failures.

nickrmc83 commented on plugins/database/rocksdb.cpp:
```diff
@@ -365,17 +397,39 @@ Status RocksDBDatabasePlugin::putBatch(const std::string& domain,
+      return Status(Status::kSuccessCode, error_string);
+    default:
+      LOG(ERROR) << "Terminal error encountered during putBatch: " << error_string;
+      Status(s.code(), error_string);
```
This is where we should schedule total failure.

nickrmc83 replied:
missing return

nickrmc83 commented on plugins/logger/aws_log_forwarder.h:
```diff
@@ -256,10 +258,20 @@ class AwsLogForwarder : public BufferedLogForwarder {
+      const auto& result_record_list = getResult(outcome);
       for (size_t i = batch.size(); i-- > 0;) {
         if (!result_record_list[i].GetErrorCode().empty()) {
+          // Log every 50th error.
```
Wrong

nickrmc83 committed:
Fix return

nickrmc83 committed:
Do not fail if schedule query fails, signal shutdown from rocksdb

nickrmc83 committed:
Limit background compactions to 1

nickrmc83 committed:
Fail after 5 error recovery begins

nickrmc83 committed:
Comment

nickrmc83 committed:
Fix hard error code

nickrmc83 committed:
Handle kNoError severity

nickrmc83 committed:
Add foreground error handling

nickrmc83 committed:
Variable naming

nickrmc83 committed:
Use increment and decrement operators on atomic types

nickrmc83 committed:
Fix foreground error counter

nickrmc83 committed:
Change error handling a little

nickrmc83 committed:
pre-increment

nickrmc83 committed:
Add missing semi-colon

nickrmc83 committed:
Restore previous background flush count

nickrmc83 committed:
Update rocksdb to version v7.10.2

nickrmc83 committed:
Add missing compilation targets for link

nickrmc83 committed:
Remove folly

nickrmc83 commented on libraries/third_party_libraries_manifest.json:
```diff
@@ -268,8 +268,8 @@
     "vendor": "facebook",
-    "version": "6.22.1",
-    "commit": "51b540921dd7495c9cf2265eb58942dad1f2ef72",
+    "version": "6.29.5",
```
should be 7.10.5

nickrmc83 committed:
Configure background flushes

nickrmc83 committed:
Reintstate paranoid_checks; enable atomic_flush

nickrmc83 committed:
Comment

nickrmc83 committed:
Disable paranoid checks

nickrmc83 committed:
Only allow a single background job to at a time.

nickrmc83 committed:
io optimisations

do not flush WAL on each write() operations; flush in a background thread every 5 seconds
do not perform file deletion in foreground client-owned threads
do not commit logs to WAL

nickrmc83 committed:
FlushWal on close

nickrmc83 committed:
Increase flush interval

nickrmc83 committed:
Reset thread pointer

nickrmc83 committed:
Print flush at warning level

nickrmc83 committed:
More tuning

nickrmc83 committed:
restrict open files to at most 256

nickrmc83 committed:
Increase maximum number of open files; disable atomic_flush

nickrmc83 committed:
Exit if flushWAL fails

nickrmc83 committed:
Set closing flag to false

nickrmc83 committed:
Remove lots of experiments

Remove experiments; increase max open files, disable WAL for all column families

nickrmc83 committed:
Increase block cache size

Increase block cache size to improve read performance, reduce eviction and IO

nickrmc83 committed:
Updates rocksdb to point at personla branch
